Add support for SQL databases with a PDO database class.

EreMaijala opened the pull request:
Includes database creation script for MySQL/MariaDB.

EreMaijala committed:
Add initial support for MySQL/MariaDB.

While MongoDB is still recommended, MySQL or MariaDB can now be used as an alternative.

EreMaijala committed:
Use iterator methods everywhere. Several fixes.

EreMaijala committed:
Remove unused and invalid database call.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Add a comment about the record foreign key.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Make sure the actual database is not connected before worker processes are forked.

- Add checks that the database is connected by current process.
- Reorder tasks and remove the reconnection mechanism in SolrUpdater.

EreMaijala committed:
Merge remote-tracking branch 'origin' into sqldb

LuomaJuha commented:
ERROR 1005 (HY000) at line 1 in file: '.../dbscripts/mysql.sql': Can't create table `recman`.`record` (errno: 150 "Foreign key constraint is incorrectly formed")
ERROR 1005 (HY000) at line 24 in file: '.../dbscripts/mysql.sql': Can't create table `recman`.`record_attrs` (errno: 150 "Foreign key constraint is incorrectly formed")
Query OK, 0 rows affected (0.033 sec)

ERROR 1005 (HY000) at line 40 in file: '.../dbscripts/mysql.sql': Can't create table `recman`.`dedup_attrs` (errno: 150 "Foreign key constraint is incorrectly formed")

Pari virhettä tuli vastaan, toimi kun ajoi toistamiseen tuon mysql.sql tiedoston

EreMaijala committed:
Fix updating of multiple records.

EreMaijala committed:
Disable foreign key checks while creating tables.

EreMaijala commented:
@LuomaJuha Now fixed.

EreMaijala committed:
Fix iteration to always sort by _id.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Make iterator faster (apart from startup). Fix $or queries.

EreMaijala committed:
Fix dedup record and collection iteration.

EreMaijala committed:
Fix iteration with an _id filter.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Fix MongoDatabase::findQueuedIds.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Fix too long table name for queue collections.

EreMaijala committed:
Clean up sample config and fix AuthorityEnrichment config.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Add mark field to record table.

EreMaijala committed:
Rework the way records are updated to handle unset values properly.

EreMaijala committed:
Add triggers to guard against dedup record being deleted while linked to.

Some reordering of database operations was necessary to not cause the triggers to fail.

EreMaijala committed:
Merge branch 'dev' into sqldb

demiankatz commented on conf/recordmanager.ini.sample:
```diff
@@ -32,18 +32,28 @@ unicode_normalization_form = NFKC
+;connection = mongodb://{username}:{password}@localhost
+;connection = mongodb://localhost
+; MongoDB database to use
 database = recman
```
If this is a Mongo-specific setting, seems like renaming it to reduce confusion might be a good idea. Or alternatively, do you have separate [Mongo] and [PDO] sections, and then pick which one is used through a  [Database] section? That might make the overall configuration less potentially confusing, especially if more backends are introduced in future.

EreMaijala replied:
Good idea. Now separated. Added bonus is that the MongoDB config is now fully backwards compatible without any tweaks.

demiankatz replied:
Great, glad I could help!

demiankatz reviewed (commented):
I finally had time to take a quick look at this. I obviously couldn't go into great depth due to lack of familiarity with the code, but I left one thought on the configuration, and I approved of what I saw in the refactoring. Do you currently use a static analysis tool of some sort? I'd strongly suggest trying PHPMD (or one of the newer ones like PHPStan or Psalm) just to make sure there are no orphaned variables/parameters after all the refactoring.

EreMaijala commented:
@demiankatz Thanks for your feedback, appreciated! No static analysis tools used yet, but I'll see what I can do about that. It's a good idea! Also, splitting the database config to backend-specific sections makes sense. I'll adjust it right away.

demiankatz commented:
Great! Let me know if I can do anything more to help. If you want to try PHPMD, you should be able to steal the Phing settings from VuFind pretty easily. You can also find Psalm setup in a PR: https://github.com/vufind-org/vufind/pull/1764. I haven't gotten around to investigating PHPStan yet.

EreMaijala committed:
Split database configuration to backend-specific sections.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Fix remaining errors reported by Psalm and enable it in Travis.

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Merge branch 'dev' into sqldb

EreMaijala committed:
Merge branch 'dev' into sqldb
